packages/core/src/shared/fs/node/fs.ts

```diff
@@ -0,0 +1,77 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import fs, { FileSystem } from '../fs'
+import vscode from 'vscode'
+import _path from 'path'
+import _fsNode from 'fs/promises'
+import lockFile, { LockOptions } from 'proper-lockfile'
+import { ToolkitError } from '../../errors'
+
+/**
+ * FileSystem related methods that only work in Node.js
+ */
+export class NodeFileSystem {
+    private fsCommon: FileSystem = fs
+
+    protected constructor() {}
+    static #instance: NodeFileSystem
+    static get instance(): NodeFileSystem {
+        return (this.#instance ??= new NodeFileSystem())
+    }
+
+    /**
+     * Acquires a lock on the given file, then runs the given callback, releasing the lock when done.
```

Comment:
is this being used to coordinat with proceses outside of vscode? 

if we are synchronizing between vscode instances, can we use globalState to coordinate instead?

Reply:
Due to us not having access to VSCode apis when the extension is being shutdown in `deactivate()`, we would not be able to update global state when needed. But since we can use the nodeFs instead, that is the next best option.

Reply:
Does that mean we could potentially lock for X seconds on deactivate...?